src/platform/packages/shared/kbn-unified-data-table/src/hooks/use_virtualization.ts

```diff
@@ -0,0 +1,135 @@
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+import { useMemo, useRef, useState, RefObject, useEffect } from 'react';
+import { type EuiDataGridProps } from '@elastic/eui';
+import { throttle } from 'lodash';
+import { DataLoadingState, DataGridPaginationMode } from '../types';
+import { useRestorableRef } from '../restorable_state';
+
+const VIRTUALIZATION_OPTIONS: EuiDataGridProps['virtualizationOptions'] = {
+  // Allowing some additional rows to be rendered outside
+  // the view minimizes pop-in when scrolling quickly
+  overscanRowCount: 20,
+};
+
+export interface UseVirtualizationProps {
+  containerRef: RefObject<HTMLSpanElement>;
+  loadingState: DataLoadingState;
+  paginationMode: DataGridPaginationMode;
+  defaultColumns: boolean;
+}
+
+export interface UseVirtualizationReturn {
+  hasScrolledToBottom: boolean;
+  virtualizationOptions: EuiDataGridProps['virtualizationOptions'];
+}
+
+export const useVirtualization = ({
+  containerRef,
+  loadingState,
+  paginationMode,
+  defaultColumns,
+}: UseVirtualizationProps): UseVirtualizationReturn => {
+  const loadingStateRef = useRef<DataLoadingState>(loadingState);
+  loadingStateRef.current = loadingState;
+  const [hasScrolledToBottom, setHasScrolledToBottom] = useState(false);
+  const hasScrolledToBottomRef = useRef<boolean>(hasScrolledToBottom);
+  hasScrolledToBottomRef.current = hasScrolledToBottom;
+  const scrollTopRef = useRestorableRef('scrollTop', 0);
+  const scrollLeftRef = useRestorableRef('scrollLeft', 0);
+  const isInitialScrollAppliedRef = useRef<boolean>(
+    !scrollTopRef.current && !scrollLeftRef.current
+  );
+
+  const virtualizationOptions: EuiDataGridProps['virtualizationOptions'] = useMemo(() => {
+    const initialScroll =
+      !isInitialScrollAppliedRef.current && (scrollTopRef.current || scrollLeftRef.current)
+        ? { top: scrollTopRef.current, left: scrollLeftRef.current }
+        : undefined;
+
+    const options: EuiDataGridProps['virtualizationOptions'] = {
+      initialScrollTop: initialScroll?.top,
+      initialScrollLeft: initialScroll?.left,
+      onScroll: throttle((event: { scrollTop: number; scrollLeft: number }) => {
+        if (loadingStateRef.current !== DataLoadingState.loaded) {
+          return;
+        }
+
+        if (isInitialScrollAppliedRef.current) {
+          scrollTopRef.current = event.scrollTop;
+          scrollLeftRef.current = event.scrollLeft;
+        } else if (event.scrollTop === 0 && event.scrollLeft === 0) {
+          // onScroll is called right after the first render
+          const rendered = Boolean(getVirtualizedElement(containerRef));
+
+          if (rendered) {
+            requestAnimationFrame(() => {
+              getVirtualizedElement(containerRef)?.scrollTo?.({
```

Comment:
Restoring the scroll position seems to work perfectly when there's a defined row height, but when I set it to auto, I encountered some really weird UI issues when switching back to the tab:

https://github.com/user-attachments/assets/202d3927-3289-4dea-82e6-2db9b21e133a

The data grid never fails to be frustrating 🙃 I'm not even sure if we can fix this on our end tbh, and maybe it's related to why `initialScrollTop` and `initialScrollLeft` don't seem to work as expected.

If we can't figure out a reliable way to restore the scroll position yet, we could split that out into a separate PR and merge what's working now if you'd like.